docs/user/exercises/programming-exercise-features.inc

```diff
@@ -87,3 +90,46 @@ Instructors can still use those templates to generate programming exercises and
 
 We encourage instructors to contribute improvements to the existing ``templates`` or to provide new templates. Please contact Stephan Krusche and/or create Pull Requests in the GitHub repository.
 
+.. _DejaGnu:
+
+Java DejaGnu: Blackbox Testing
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The classical testing frameworks, like JUnit for example, allows writing white box tests.
+This enforces many assumptions about the structure of the source code under test,
+like the class and methods names and even the exact method signatures.
+This implies for a Programming Exercise you have to specify all the structural aspect
+if your test suite should be able to run on the student submissions.
+That is maybe okay or even desired for a beginner course,
+but for advanced courses this is a downside,
+because without making their own decision about the structure of their source code,
+the students cannot collect experience in this important aspect of programming through your exercises.
+
+To be able to include the structural aspect of programming in your Programming Exercises,
+without losing the ability to test submissions automatically, you can use DejaGnu.
+It allows you to write blackbox tests for command line interfaces of applications.
+The tests are written in Expect Script, an extension to the Tcl scripting language.
+Expect is a traditional Unix utility for automatic interactions with programs that expose a text terminal
+interface in a robust way.
+The test script starts the program to test as own process (possibly several times) and interactively controls
+it by textual inputs via pipelining.
+Using the respective outputs, the script evaluates the functionality and reacts accordingly in a
+fine-grained manner.
+Just like a human user would test the running program manually.
```

Comment:
```suggestion
The script then reads the respective outputs and can make 'assertions' (e.g. exact or regex matching) on them.
Based on the obtained output it can also decide on the next inputs, thereby simulating how a user might interact with the program.
```
I think ‘fine-grained manner’ is a bit vague. The suggestion gives a less abstract example what the tool can do.